### PR TITLE
[New] Set up location-driven geotriggers

### PR DESCRIPTION
**Pull request by CalebRas**
## Description

This PR implements `Set up location-driven geotriggers` in `Display information` category.
URL to README: https://github.com/Esri/arcgis-maps-sdk-swift-samples/tree/Caleb/New-SetUpLocationDrivenGeotrigers/Shared/Samples/Set%20up%20location-driven%20geotriggers

## Linked Issue(s)

- `swift/issues/4135`

## How To Test

* Watch the location display move and the status texts change accordingly.
* Tap the "Current Section" button to see the current section information when available.
* Tap the "Point of Interest" button to see the point of interest information when available.

## Screenshots

set-up-location-driven-geotriggers

### CODE REVIEW

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:308]:
```suggestion
```
This is unnecessary. `.task(id: isShowingPopup)` will start the location display when the view is loaded.

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:298]: Can you make this method `private static func makeLocationDisplay(dataSource: LocationDataSource) -> LocationDisplay` and update its docstring?

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:115]:
```suggestion
        let locationDisplay: LocationDisplay
```
Then change it here.

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:178]: Since the end result we want is a polyline, put it on the `Polyline` extension at the bottom to declutter the model.

```swift
private extension Polyline {
    /// A path in Santa Barbara Botanic Garden.
    static var walkingTour: Polyline {
        let walkingTourPolylineJSON = ...
        let routePolyline = try? Polyline.fromJSON(walkingTourPolylineJSON)
        let densifiedRoute = GeometryEngine.geodeticDensify(
            routePolyline!,
            maxSegmentLength: 5.0,
            lengthUnit: .meters,
            curveType: .geodesic
        ) as! Polyline
        return densifiedRoute
    }
}
```

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:179]: And the initializer will become

```swift
        init() {
            // Create simulated data source with the route polyline.
            locationDataSource = SimulatedLocationDataSource(polyline: .walkingTour)

            locationDisplay = Self.makeLocationDisplay(dataSource: locationDataSource)
        }
```

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:50]: I'm confused by the manual negative padding here. 😵‍💫 A linebreak is added in the first Text, then a negative padding is used to put the second text closer to the top… Is it intended?

Occasionally manually adjust the padding to make the view look better is OK, but it is generally an anti-pattern for adaptive layout. Try to avoid them as much as we can.

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:60]: The old sample supports displaying multiple popups in a page view-like controller. Can you explore if that is possible in [SwiftUI](https://developer.apple.com/documentation/swiftui/tabviewstyle/page)? It's OK if it doesn't.

@mhdostal I vaguely recall we won't support multi-page popup in Swift toolkit, is that correct?

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:98]: Very unlikely to throw an error as the location data source is just a string, so I would use `try?` to omit the error here.

File: Shared/Samples/Set up location-driven geotriggers/README.md
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/README.md:17]:
```suggestion
1. Create a `GeotriggerFeed` with a `LocationDataSource` object (in this case, a `SimulatedLocationDataSource`).
```

**Review by yo1995 — COMMENTED** [review]
First batch of comments.

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (CalebRas) [reply Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:59]: I looked into and don't see an easy way to do it. Especially if it isn't supported in the toolkit.

Comment (mhdostal) [reply Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:59]: > I vaguely recall we won't support multi-page popup in Swift toolkit, is that correct?

Correct, multi-page popups are not supported in the popup in the Swift Toolkit.

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:199]: A few improvements to this method:

1. `makeGeotrigger` implies it is a [factory method](https://www.swift.org/documentation/api-design-guidelines#strive-for-fluent-usage). But this method is not returning anything. A possible change is `makeGeotriggerMonitor` and let it return a monitor object. Then in the outer driver method (`startGeotriggerMonitoring`, which may have a better name), add the created monitor objects to the array.
2. Starting the monitors can also be done in the driver method, so this method is free from async context.
3. Currently there is only 1 `notificationTaskHandle` in the data model, but multiple async sequences are monitored. We can discuss this more thoroughly. Please keep an eye on the discussion in the swift-general channel.
    - Use a Task block to wrap a task group of Void.self, then add all the sequences to the group
    - Use 1 Task block for each sequence, and store the block handles in an array
    - Use multiple task modifiers

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:242]: Though for this sample's data it is safe to force unwrap, I recommend `guard let` at the top in case the users might have different data.

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:246]: Interweaving Swift concurrency and Dispatch (GCD) is typically not a good idea. The only thing that needs to be on the main queue is the UI code, and other code can be on arbitrary threads.

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (yo1995) [inline Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:271]: Put this method on the main actor. It would trigger a series of async context errors (so-called "async/await hell" in JavaScript). Try to restructure the logic. Leave it to me if you have troubles with it.
```suggestion
        @MainActor
        private func updateStatusText(featureName: String, notificationType: FenceGeotriggerNotificationInfo.NotificationType) {
```

**Review by yo1995 — COMMENTED** [review]
Getting into the concurrency deep water! 🎣

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (CalebRas) [reply Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:59]: It works but it super glitchy. The popup kind of spazzes when and wont let you all the way over to the other screen when you try to swipe.

File: Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift
Comment (CalebRas) [reply Shared/Samples/Set up location-driven geotriggers/SetUpLocationDrivenGeotriggersView.swift:260]: I am not totally sure how to fix this without throwing publishing changes from background threads errors.

**Review by yo1995 — COMMENTED** [review]
Here are some docs changes. I'll add more feedback on the data model soon.

**Review by yo1995 — COMMENTED** [review]
Instead of making separate comments, I decided to stash my suggestions on a [branch](https://github.com/Esri/arcgis-maps-sdk-swift-samples/tree/Ting/GeotriggerSuggestions/Shared/Samples/Set%20up%20location-driven%20geotriggers). Here are the compared branches: https://github.com/Esri/arcgis-maps-sdk-swift-samples/compare/Caleb/New-SetUpLocationDrivenGeotrigers...Ting/GeotriggerSuggestions?expand=1

We can discuss these suggestions in detail. Once we are all good, you can merge it into your branch locally and update the PR. Or if you want me to comment on the changes on GitHub, feel free to open a PR targeting your branch and I can add the comments on their lines.

**Review by yo1995 — APPROVED** [review]

**Comment by yo1995** [comment]
Hi @mhdostal if you have a chance can you review this PR? Or assign to someone else using the sample reviewers alias.

**Review by mhdostal — APPROVED** [review]
Looks great, very nice work!

**Pull request merged**
